Improve wizard hardware detection

halfline opened the pull request:
This set of commits should make the setup wizard stop promoting the small model so aggressively.  I've only tested on linux, I will need to test on mac before we can merge.

halfline committed:
Include GPU VRAM field to SystemInfo

At the moment there is no way to query the amount of video memory
a gpu has, but for discrete cards, that's the actually interesting
thing to know.

This commit extends the GPU information returned in SystemInfo by
adding a 'vram' field to the GPU objects.

This commit also updates getSystemGPUs() accordingly.

halfline committed:
Add new shouldRecommendLargeModel function

This commit adds a successor for the isHighEndMachine
function called shouldRecommendLargeModel that takes
into account the video memory of the machine on systems
with discrete graphics cards.

A future commit will use this function to improve the
setup wizard.

halfline committed:
Improve wizard's hardware-based recommendations

At the moment the setup wizard only looks at system
memory and platform to decide if the large model or
small model is appropriate.

It also hardcodes a 32GB limit which is higher than
actually needed to get good performance on systems
that keep the model in system memory (apple hardware).

But for non-apple harware, machines with discrete
graphics cards, the amount of video memory is what
really matters.

Also, the small model doesn't work that well for real
world developing, so we shouldn't really ever recommend it.

This commit:

1. Changes the wizard to look at video memory on non-apple
hardware

2. Changes the wizard to look at the actual model memory
recommendations, instead of hardcoding 32GB.

3. Gets decidedly less enthusiastic about recommending the
small model to the user. It replaces the recommendation with
a warning.

4. Renames recommendedModel to preselectedModel to make the
code more align with the new messaging.

halfline commented:
so trying it on mac it seems to work as designed, but looking closer my design tweak is a little wacky:

![image](https://github.com/user-attachments/assets/e956c7ea-3e80-442a-b0b9-fdbdf08f9ffd)

It recommends the small model for "less than 8 gb" and the large model for "more than 16 gb". So all those users with 12gb cards won't know what to do. probably just need to pick the bigger threshold.

halfline commented:
the third commit message was improperly squashed as well. i need to fix that.

halfline commented:
turns out IsHighEndMachine is used elsewhere in the code, so i dropped the commit that dropped that definition.

owtaylor commented on core/granite/commons/sysInfo.ts:
```diff
@@ -69,3 +71,19 @@ export function getRecommendedModels(systemInfo: SystemInfo) {
+    return Math.max(max, (gpu.vram || 0) * GB);
+  }, 0);
+
+  return maxVRAM >= requirements.recommendedMemoryBytes;
```
I don't think we should be using the same number for the VRAM threshold for a discrete GPU and the memory requirement for a Mac (or eventually other iGPU systems). 12GB of VRAM for a discrete GPU is quite reasonale for the :8b model. 12GB of RAM is not.

halfline replied:
okay will add a 6GB offset for other tasks on system when using system memory

halfline replied:
so truthfully the 8b model takes like 8gb of memory and we have it set like:

```
    minMemoryBytes: 12 * GB,
    recommendedMemoryBytes: 16 * GB,
```
So there's already a buffer built in that both discrete and mac hardware cases use, but:

- we only use recommendedMemoryBytes for both, and the buffer it has there isn't really right for either. I think we should drop minMemoryBytes and make recommendedMemoryBytes represent the actual usage of the model in memory, then have two model agnostic "reserve" numbers for amounts of the allocation of system and graphics memory that we won't encroach on when we do model selection.

owtaylor reviewed (changes requested):
Other than the threshold issue, looks good to me.

halfline committed:
modelRequirements: Drop minMemoryBytes

Each model has an object of hardware requirements it
ostensibly requires for inference to function reasonably.

That object has a "minMemoryBytes" field that is only
used by an unused function, checkModelCompatibility.

This commit drops both.

halfline committed:
modelRequirements: Use different requirements for cpu vs. gpu

If the model is hosted in system memory, like on Apple hardware,
it can't really use all available ram for itself, since the OS, Safari,
the application getting written, etc, take up memory as well.

That is true to some extent for graphics card memory as well, but to a
far lesser degree.

The difference is exacerbated by the fact that ollama, itself, will be
more greedy about its own memory needs if it detects its running in
system ram, but that's an issue to address later.

This commit unfurls that reserved memory value from the model
requirements, and splits it out into two distinct values, one that's
used for available system memory calculations and one that's used for
graphics card memory calculations.

halfline committed:
Granite.Code: Drop context length to 8k

Using 16k context constrains the available hardware we're able
to run on a little too much.

At some point, we need to dynamically dial in context list to
be optimal for the system running the model, but for now,
this commit drops the context length to 8k.

halfline commented:
I think this is probably good to go now, but I want to check it on apple hardware before merging. Owen do you want to take a look at the new commits?

halfline commented:
just did some quick testing on the mac mini with the updates and it seems to still properly recommend the large model.

halfline commented on core/granite/commons/systemRequirements.ts:
```diff
@@ -0,0 +1,11 @@
+
+export const SYSTEM_REQUIREMENTS: SystemRequirements = {
+  reservedSystemMemory: 6 * GB,
+  reservedGraphicsMemory: 0 * GB,
```
Note i set this to 0 for now so medium end discrete cards still recommend the large model. we may need to tweak this value.
